File: 7.go
package main

import (
	"bufio"
	"fmt"
	"os"
)

type file struct {
	name string
	size int
}

type dir struct {
	name   string
	parent *dir
	files  []file
	dirs   []*dir
}

func (d *dir) size() int {
	result := 0
	for _, file := range d.files {
		result = result + file.size
	}
	for _, dir := range d.dirs {
		result = result + dir.size()
	}
	return result
}

func (d *dir) total() int {
	result := 0
	if d.size() <= 100000 {
		result = result + d.size()
	}
	for _, dir := range d.dirs {
		result = result + dir.total()
	}
	return result
}

func (d *dir) smallest(threshold int) int {
	result := 10000000000
	for _, dir := range d.dirs {
		size := dir.smallest(threshold)
		if size >= threshold && size < result {
			result = size
		}
	}
	size := d.size()
	if size >= threshold && size < result {
		result = size
	}
	return result
}

func main() {
	scanner := bufio.NewScanner(os.Stdin)
	root := dir{name: "/", files: []file{}, dirs: []*dir{}}
	current := &root
	for scanner.Scan() {
		line := scanner.Text()
		if line[:4] == "$ ls" {
			// skip
		} else if line[:5] == "$ cd " {
			switch line[5:] {
			case "/":
				current = &root
			case "..":
				current = current.parent
			default:
				for _, dir := range current.dirs {
					if dir.name == line[5:] {
						current = dir
					}
				}
			}
		} else if line[:3] == "dir" {
			current.dirs = append(current.dirs, &dir{name: line[4:], files: []file{}, dirs: []*dir{}, parent: current})
		} else {
			var file file
			fmt.Sscanf(line, "%d %s", &file.size, &file.name)
			current.files = append(current.files, file)
		}
	}
	fmt.Println((&root).total(), (&root).smallest(30000000-(70000000-(&root).size())))
}

File: 7.swift
import Foundation

class Directory {
    var subDirs: [String: Directory] = [:]
    var files: [String: Int] = [:]
    var parent: Directory?
    var size: Int {
        return files.values.reduce(0, +) + subDirs.values.map { $0.size }.reduce(0, +)
    }
    var allSizes: [Int] {
        var sizes: [Int] = []
        self.forEach { sizes.append($0.size) }
        return sizes
    }

    init() { }
    init(_ p: Directory) { parent = p }
    func forEach(_ f: (Directory) -> Void) {
        f(self)
        subDirs.forEach { $0.value.forEach(f) }
    }
}

var root = Directory()
var current = root
while let line = readLine() {
    let words = line.components(separatedBy: [" "])
    if words[0] == "$" { 
        if words[1] == "cd" {
            switch words[2] {
                case "/":  current = root
                case "..": current = current.parent!
                default:   current = current.subDirs[words[2]]!
            }
        }
    } else if words[0] == "dir" {
        current.subDirs[words[1]] = Directory(current)
    } else {
        current.files[words[1]] = Int(words[0])!
    }
}
print(root.allSizes.filter {$0 <= 100000}.reduce(0, +), 
      root.allSizes.filter {$0 >= 30000000-(70000000-root.size)}.min() ?? 0)